modules/spring-config/src/main/java/org/mule/config/spring/SpringRegistry.java

```diff
@@ -203,48 +334,112 @@ public Object lookupObject(String key)
+            }
 
-    @Override
-    public void registerObjects(Map<String, Object> objects) throws RegistrationException
-    {
-        throw new UnsupportedOperationException("Registry is read-only so objects cannot be registered or unregistered.");
-    }
+            return dependents;
+        }
 
-    @Override
-    public void unregisterObject(String key)
-    {
-        throw new UnsupportedOperationException("Registry is read-only so objects cannot be registered or unregistered.");
+        throw new UnsupportedOperationException("This operation is only available when this registry is backed by a ConfigurableApplicationContext");
     }
 
-    @Override
-    public void unregisterObject(String key, Object metadata) throws RegistrationException
+    private class ConfigurableRegistrationDelegate implements RegistrationDelegate
     {
-        throw new UnsupportedOperationException("Registry is read-only so objects cannot be registered or unregistered.");
+
+        private final ConfigurableApplicationContext applicationContext;
+
+        private ConfigurableRegistrationDelegate(ConfigurableApplicationContext applicationContext)
+        {
+            this.applicationContext = applicationContext;
+        }
+
+        @Override
+        public void registerObject(String key, Object value) throws RegistrationException
+        {
+            doRegisterObject(key, value);
+        }
+
+        @Override
+        public void registerObject(String key, Object value, Object metadata) throws RegistrationException
+        {
+            registerObject(key, value);
+        }
+
+        @Override
+        public void registerObjects(Map<String, Object> objects) throws RegistrationException
+        {
+            if (objects == null || objects.isEmpty())
+            {
+                return;
+            }
+
+            for (Map.Entry<String, Object> entry : objects.entrySet())
+            {
+                registerObject(entry.getKey(), entry.getValue());
+            }
+        }
+
+        @Override
+        public Object unregisterObject(String key) throws RegistrationException
+        {
+            Object object = applicationContext.getBean(key);
+
+            if (applicationContext.getBeanFactory().containsBeanDefinition(key))
+            {
+                ((BeanDefinitionRegistry) applicationContext.getBeanFactory()).removeBeanDefinition(key);
+            }
+
+            ((DefaultListableBeanFactory) applicationContext.getBeanFactory()).destroySingleton(key);
+
+            return object;
+        }
+
+        private synchronized void doRegisterObject(String key, Object value) throws RegistrationException
+        {
+            if (applicationContext.containsBean(key))
+            {
+                if (logger.isWarnEnabled())
+                {
+                    logger.warn(String.format("Spring registry already contains an object named '%s'.  he previous object will be overwritten.", key));
```

Comment:
"he previous object will be overwritten" -> "The previous object will be overwritten"